### ISSUE TITLE
Javascript in components

### ISSUE THREAD

**Issue by lucosius** [putyourlightson/craft-sprig#227]
My again :(

I cannot figure out the way to make Javascript to work in Sprig. I have tabs and some JS for the images gallery. It looks like this:

```
{% set entry = craft.entries()
    .id(entryId)
    .one() %}
{% set info = info ?? 'specs' %}
{{ hiddenInput('info', info) }}
<section class="section pt-2 pb-0">
  <div class="container">
        <div class="columns tabs is-multiline is-mobile has-border-bottom pb-2 pt-2 mb-0">
                        <div class="column is-2-desktop is-offset-4-desktop is-12-mobile is-12-tablet has-text-right">
     <a href="{{ entry.url }}/sprendimai"><button class="button is-evioni-buttons is-success has-text-primary is-size-6 is-fullwidth"><span class="light-icon">Sprendimas</span></button></a>
   </div>
        <div sprig s-val:info="specs" s-val:reset="1" class="column is-2-desktop is-offset-1-desktop is-6-mobile is-6-tablet"><button class="button is-evioni-rounded is-evioni-shaddow is-fullwidth has-background-transparent {{ info == 'specs' ? 'is-active' }}">Specifikacijos</button></div>
     <div sprig s-val:info="gallery" s-val:reset="1" class="column is-2-desktop is-offset-1-desktop is-6-mobile is-6-tablet"><button class="button is-evioni-rounded is-evioni-shaddow is-fullwidth has-background-transparent {{ info == 'gallery' ? 'is-active' }}">Galerija</button></div>
</div>
  <div>
      </div>
</div>
</section>

{% if info == 'specs' %}
some content here
  {% endif %}

{% if info == 'gallery' %}
 {{ sprig('database/gallery.twig', {'entryId': entryId}) }}
{% endif %}

<script src='https://cdn.jsdelivr.net/npm/jquery@3.6.0/dist/jquery.js'></script>
<script src='https://cdn.jsdelivr.net/npm/lightgallery@2.0.0-beta.3/lightgallery.umd.js'></script>
<script src='https://cdn.jsdelivr.net/npm/lightgallery@2.0.0-beta.3/plugins/zoom/lg-zoom.umd.js'></script>
<script src='https://cdn.jsdelivr.net/npm/justifiedGallery@3.8.1/dist/js/jquery.justifiedGallery.js'></script>
<script src='https://cdn.jsdelivr.net/npm/lightgallery@2.0.0-beta.3/plugins/thumbnail/lg-thumbnail.umd.js'></script>
<script>
const $galleryContainer = document.getElementById("gallery-container");

const customButtons = `<button type="button" id="lg-toolbar-prev" aria-label="Previous slide" class="lg-toolbar-prev lg-icon">  </button><button type="button" id="lg-toolbar-next" aria-label="Next slide" class="lg-toolbar-next lg-icon">  </button>`;

$galleryContainer.addEventListener("lgInit", (event) => {
  const pluginInstance = event.detail.instance;

  // Note append and find are not jQuery methods
  // These are utility methods provided by lightGallery
  const $toolbar = pluginInstance.outer.find(".lg-toolbar");
  $toolbar.prepend(customButtons);
  document.getElementById("lg-toolbar-prev").addEventListener("click", () => {
    pluginInstance.goToPrevSlide();
  });
  document.getElementById("lg-toolbar-next").addEventListener("click", () => {
    pluginInstance.goToNextSlide();
  });
});

lightGallery($galleryContainer, {
  speed: 500,
  zoomPluginStrings: true,
  getCaptionFromTitleOrAlt: false,
  download: false,
  controls: true,
  zoom: false,
  plugins: [lgZoom]
});

</script>

```

If I set default tab to the gallery - everything is working fine but otherwise - not. I tried to move JS to gallery.twig file but it didn't help. What I'm doing wrong here?

**Comment by bencroker** [comment]
The general best practice is to put all source JS scripts (`<script src="">`) into the main layout template and all JS code (`<script>...</script>`) inside the Sprig component.

See https://putyourlightson.com/plugins/sprig#javascript

**Comment by lucosius** [comment]
Honestly, I don't really understand what I need to do. If I move source to the footer of website and keep JS code inside sprig - it doesn't work at all. I guess it's because the code is above the source scripts?

In the way I had scripts - it was working until the other tab is clicked. I checked the link that you gave like million times and not sure what solution should I use to make it work.

Demo is here:

https://dev.telefonai.eu/elektromobiliai/mercedes-eqa-250-2

On desktop version, where gallery is default tab, it's working fine. On mobile version, where gallery is secondary tab, it doesn't work.

**Comment by bencroker** [comment]
> If I move source to the footer of website and keep JS code inside sprig - it doesn't work at all. I guess it's because the code is above the source scripts?

That's probably the issue, yes. I would solve it by defining a function and calling it at the appropriate time.

```twig
{# Layout #}

<script src="..."></script>
<script src="..."></script>
<script src="..."></script>
<script>
function setupGallery() {
    ...
}
setupGallery();
</script>
```

```twig
{# Sprig component #}

{% if sprig.isRequest %}
    <script>
        setupGallery();
    </script>
{% endif %}
```

**Comment by lucosius** [comment]
Now it's better but it's still doesn't work in secondary tab. It looks like I cannot find the proper place to put setupGallery function in the component. Basically it works great until I switch tabs.

**Comment by bencroker** [comment]
Great, so you just need to figure out where to place the code now!

**Comment by lucosius** [comment]
hmm. I tried everything but it just doesn't work if i make a switch between the tabs.

I have:

Main layout with JS codes and setupGallery function

Entry with:

`{{ sprig('mobile/database/info.twig', {'_info': 'specs, gallery', 'entryId': entry.id}) }}`

Info template with the code that I put in the first post. I tried to put in all of these files but it doesn't work as it should. As I said, it works only with the condition that gallery is in the first tab. If it's in any other tab or you switched between the tabs once - it doesn't work anymore.
